Add comprehensive API features: Swagger docs, throttling, JWT auth, a…

jevgenius opened the pull request:
…nd tests

- Configure drf-spectacular for OpenAPI/Swagger documentation
- Add detailed parameter descriptions for movie and session filtering
- Implement rate limiting: 10 req/min for anon, 30 req/min for auth users
- Add JWT authentication with token/refresh endpoints
- Create comprehensive MovieViewSet test suite (18 tests)
- Cover CRUD operations, filtering, permissions, and edge cases
- Fix permissions to allow unauthenticated read access
- Clean up unused imports

All API endpoints now have proper documentation and throttling protection.

jevgenius committed:
Add comprehensive API features: Swagger docs, throttling, JWT auth, and tests

- Configure drf-spectacular for OpenAPI/Swagger documentation
- Add detailed parameter descriptions for movie and session filtering
- Implement rate limiting: 10 req/min for anon, 30 req/min for auth users
- Add JWT authentication with token/refresh endpoints
- Create comprehensive MovieViewSet test suite (18 tests)
- Cover CRUD operations, filtering, permissions, and edge cases
- Fix permissions to allow unauthenticated read access
- Clean up unused imports

All API endpoints now have proper documentation and throttling protection.

mateacademy-ai-mentor commented on cinema/permissions.py:
```diff
@@ -4,10 +4,6 @@
-                and request.user
-                and request.user.is_authenticated
-            )
+            request.method in SAFE_METHODS
```
Issue: The current logic allows any user (authenticated or not) to perform safe (read-only) methods. However, the class name suggests that only authenticated users should have read-only access, and only admin users should have write access. You should check that the user is authenticated for safe methods as well.

mateacademy-ai-mentor commented on cinema/tests/test_movie_api.py:
```diff
@@ -66,6 +68,234 @@ def detail_url(movie_id):
+    
+    def test_filter_movies_by_title(self):
+        """Test filtering movies by title"""
+        response = self.client.get(MOVIE_URL, {"title": "Movie 1"})
```
Issue: The filter value for 'title' is 'Movie 1', but the created movie titles are 'Test Movie 1' and 'Test Movie 2'. This will not match any movies and the test will not behave as intended. Please update the filter value to match the actual movie title, e.g., 'Test Movie 1'.

mateacademy-ai-mentor commented on cinema/tests/test_movie_api.py:
```diff
@@ -66,6 +68,234 @@
+    
+    def test_filter_movies_by_title_case_insensitive(self):
+        """Test filtering movies by title is case insensitive"""
+        response = self.client.get(MOVIE_URL, {"title": "movie 1"})
```
Issue: The filter value for 'title' is 'movie 1', but the created movie titles are 'Test Movie 1' and 'Test Movie 2'. This will not match any movies and the test will not behave as intended. Please update the filter value to match the actual movie title, e.g., 'test movie 1'.

mateacademy-ai-mentor commented on cinema/tests/test_movie_api.py:
```diff
@@ -66,6 +68,234 @@
+    def test_filter_movies_combined_filters(self):
+        """Test filtering movies with multiple filters"""
+        response = self.client.get(MOVIE_URL, {
+            "title": "Movie 1",
```
Issue: The filter value for 'title' is 'Movie 1', but the created movie titles are 'Test Movie 1' and 'Test Movie 2'. This will not match any movies and the test will not behave as intended. Please update the filter value to match the actual movie title, e.g., 'Test Movie 1'.

mateacademy-ai-mentor reviewed (changes requested):
### Review Summary

Thank you for your hard work! There are a couple of critical issues that need to be addressed before your PR can be approved. Most importantly, the filtering test for movies by title uses a value ('Movie 1') that does not match the created test data ('Test Movie 1', 'Test Movie 2'), so the test will not properly verify the filtering functionality. Please update the filter value to match the actual movie title, such as 'Test Movie 1'. Additionally, the custom permission logic currently allows any user (even unauthenticated) to perform safe (read-only) methods, but the class name and requirements indicate that only authenticated users should have read-only access, and only admin users should have write access. Please update the permission logic to enforce this. Once these issues are fixed, your solution will meet all requirements! Keep going, you're almost there! 🚀

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)

> [!IMPORTANT]
> Fix the issues, commit the changes, and then re-request my review using the **`Re-request review` button**[^2] in the Reviewers tab.

[^2]: [Re-request AI review example](https://snipboard.io/nOXxHN.jpg)

> [!TIP]
> If you have any comments about this review, please leave a message in the [Add a comment](#new_comment_field) section below and **mention me with the `@mateacademy-ai-mentor` tag**[^3].
> I will read your message during my next review of your pull request. Please do not reply directly to my comments, as that feature is not supported yet.

[^3]: [Add AI review comments example](https://snipboard.io/hVlOey.jpg)

> [!CAUTION]
> Do not close the pull request until the review is complete. Closing it early can lead to unexpected behavior and disrupt the review process.

jevgenius committed:
Fix critical issues: correct test data and enforce authentication for read access

- Fix movie filtering tests to use correct test data ('Test Movie 1' instead of 'Movie 1')
- Update permission logic to require authentication for read-only operations
- Ensure IsAdminOrIfAuthenticatedReadOnly enforces proper access control:
  - Only authenticated users can read (GET) movies
  - Only admin users can create/modify movies
- Update all relevant tests to authenticate users for read operations
- All 18 tests now pass correctly

mateacademy-ai-mentor reviewed (approved):
### Review Summary

Great job! 🎉 Your code fully meets all the core requirements and constraints of the task. All previous issues have been resolved, your tests are comprehensive, permissions and throttling are correctly configured, and the documentation is clear and complete. Keep up the excellent work!

***

✨ Please rate this review in the Platform's notification center[^1]. I will be grateful for your feedback! ✨

[^1]: [Rate AI review example](https://snipboard.io/gD6XKF.jpg)
